File: server/src/battle/model/shootableMonster/ninja/NinjaAbility.java
package battle.model.shootableMonster.ninja;

import battle.model.shootableMonster.Ability;
import battle.model.shootableMonster.Monster;
import battle.model.shootableMonster.config.NinjaConfig;
import config.battle.BattleConfig;
import model.battle.map.Position;


public class NinjaAbility extends Ability {

    boolean status = false;
    //danh dau toa do bat dau kich hoat ki nang
    Position positionActive;

    public NinjaAbility(Monster owner) {

        super(owner);
    }

    public boolean canActive(double maxCurHp, double curHp){

        if(curHp <= maxCurHp - this.owner.getBaseHp() * NinjaConfig.RATIO_LOST_HP && !this.owner.isFrozen() && !this.owner.isStunned() && !(this.owner.speedThrown > 0)){
            return true;
        }
        return false;
    }

    @Override
    public void active(){

        this.positionActive = new Position(this.owner.position.getX(), this.owner.position.getY());
        this.status = true;
    }

    @Override
    public boolean isActive(){

        if(this.status){
            double dx = Math.abs(this.positionActive.getX() - this.owner.position.getX());
            double dy = Math.abs(this.positionActive.getY() - this.owner.position.getY());
            if( dx + dy < 3 * BattleConfig.SQUARE_SIZE){
                return true;
            }
        }
        this.status = false;
        return false;
    }

}

File: client/src/modules/base/shootableMonster/monsterNinja/NinjaAbility.js
let NinjaAbility = Ability.extend({
    ctor: function (owner){
        this._super(owner);
        this.status = false;
    },
    canActive: function (maxHp, curHp){
        if(curHp <= maxHp - this.owner.baseHp * NinjaAbilityConfig.RATIO_LOST_HP && !this.owner.isFrozen() && !this.owner.isStunned() && !(this.owner.speedThrown > 0)){
            return true;
        }
        return false;
    },
    active: function (){
        this.positionActive = cc.p(this.owner.position.x, this.owner.position.y);
        this.status = true;
    },
    isActive: function (){
        if(this.status){
            let dx = Math.abs(this.positionActive.x - this.owner.position.x);
            let dy = Math.abs(this.positionActive.y - this.owner.position.y);
            if(dx + dy < BATTLE.SQUARE_SIZE * NinjaAbilityConfig.DISTANCE){
                return true;
            }
        }
        this.status = false;
        return false;
    }

})